### DISCUSSION TITLE
Code-Generator project (NumFOCUS)

### DISCUSSION THREAD

**Discussion by vfdev-5** [pytorch/ignite#1694] [Ideas]
## Context

There is a great tool : https://traingenerator.jrieke.com/ which can help to quickly start to work on a specific task without rewriting everything: Kaggle competition, client prototype project, etc.

We would like to create a similar generator for Ignite only for daily usage with our specifications

## What we would like to do ?

- Create new repository on our organization: https://github.com/pytorch-ignite
  - Code based on https://github.com/jrieke/traingenerator
  - Change graphics to pytorch-ignite
- Host the application on Streamlit directly (already applied to Streamlit sharing)
- Generate code following our specifications:

Goals:

```
- Multiflow/multiple files application:
  - Vision
    - Classification task (similar to our [cifar10 example](https://github.com/pytorch/ignite/tree/master/examples/contrib/cifar10))
    - Object detection task
    - Image segmentation task
  - NLP
    - Text classification task (similar to our [transformers](https://github.com/pytorch/ignite/tree/master/examples/contrib/transformers))
     - ConvAI (maybe based on updated https://github.com/pytorch-ignite/transfer-learning-conv-ai)
- Configurable parts:
  - Datasets with dataflow file: e.g. get_dataloaders function
    - 1 public dataset as initial option
    - ImageDataset for vision
    - ...
  - Models via models file: get_model function
    - Models from torchvision
  - Configuration setting
    - Google-fire
    - Hydra ?
    - [py_config_runner](https://github.com/vfdev-5/py_config_runner) ?
  - Experiment tracking system support
    - Tensorboard
    - ClearML
    - Neptune/MLflow ?
- Output:
  - Local download of multiple files (as an archive?)
```

## How to participate ?

We are looking for 1 or 2 people to lead and implement this project. In this role, you will be asked to
- elaborate in detail our goals and todo list
- work on the code of the application
- deploy and maintain the application, including bug fixes, help with questions, reviewing the PRs.
- write a blog post to communicate about the application

The work will be realized in the following repository: https://github.com/pytorch-ignite/code-generator (will be online soon).

We don’t require specific qualifications or experience for this role, but the following qualities are what we’re looking for:
- fluent with Python and git
- basic knowledge of how it works with GitHub: Issues, PRs, Code of Conduct etc
- (preferred) already contributed to PyTorch-Ignite
- (preferred) already created apps with Streamlit

We encourage everyone to apply if you think you may be a good fit, even if you are unsure whether you meet all of the criteria. If so, please send an email to  contact@pytorch-ignite.ai with a brief description of your experience related to the qualities we are looking for and your motivations for the role.

For other contributions, please, simply pick a task from the [todo list](https://github.com/pytorch-ignite/code-generator/issues), say that you would like to work on that, discuss with project leads and finally, send a PR.

## Mentors

The following folks could be mentors for the project leader(s):

- Victor (@vfdev-5)
- Johannes (@jrieke): available to answer questions related to Traingenerator / Streamlit.

## Expected outcome and deadlines

We expect to launch v0.1 of the application in the mid-March 2021 and constantly improve it adding new features from the goals. Blog post about usable version could be expected to be published in the beginning of May.

## Compensation

This project is a part of NumFOCUS Small Development Grant and can be eligible for compensation on realized work. The following roles will be rewarded:
- project leaders
- major contributors (will be defined with project leaders)

Compensation can be received via:
- PayPal
- Stripe Connect

**Comment by vfdev-5** [comment]
Feel free to ask questions here

**Reply by jeffydc** [reply]
Hi @vfdev-5 ,
How long do we have to be project leaders ? (til v0.1.0 or til usable version)

**Reply by vfdev-5** [reply]
Thanks for the question @ydcjeff ! We expect project leaders to work until the usable version and the blog post is published.

**Comment by sparkingdark** [comment]
Hey, @vfdev-5 this project seems interesting can I join on it?

**Reply by vfdev-5** [reply]
Hey @sparkingdark, sure you can contribute to the code source once it is online.

**Comment by Rajathbharadwaj** [comment]
Hey I'm interested. I've sent a mail.

**Reply by vfdev-5** [reply]
Hey @Rajathbharadwaj we have an issue with your email server now, could you please resend your email to  vfdev.5 at gmail dot com. Thanks!

**Reply by Rajathbharadwaj** [reply]
Oh okay. I will resend.

**Reply by Rajathbharadwaj** [reply]
Hey @vfdev-5 please check out. I've sent a mail to vfdev dot 5 gmail dot com
